Allow commands to return a Task<RuntimeResult>

FiniteReality opened the pull request:
This allows bot developers to centralize command result logic by using result data whether the command was successful or not.

Example usage:
```csharp
var _result = await Commands.ExecuteAsync(context, argPos);
if (_result is RuntimeResult result)
{
    await message.Channel.SendMessageAsync(result.Reason);
}
else if (!_result.IsSuccess)
{
    // Previous error handling
}
```

The RuntimeResult class can be subclassed too, for example:
```csharp
var _result = await Commands.ExecuteAsync(context, argPos);
if (_result is MySubclassedResult result)
{
    var builder = new EmbedBuilder();
    for (var pair in result.Data)
    {
        builder.AddField(pair.Key, pair.Value, true);
    }
    await message.Channel.SendMessageAsync("", embed: builder);
}
else if (_result is RuntimeResult result)
{
    await message.Channel.SendMessageAsync(result.Reason);
}
else if (!_result.IsSuccess)
{
    // Previous error handling
}
```

FiniteReality committed:
Allow commands to return a Task<RuntimeResult>

This allows bot developers to centralize command result logic by
using result data whether the command as successful or not.

Example usage:
```csharp
var _result = await Commands.ExecuteAsync(context, argPos);
if (_result is RuntimeResult result)
{
    await message.Channel.SendMessageAsync(result.Reason);
}
else if (!_result.IsSuccess)
{
    // Previous error handling
}
```

The RuntimeResult class can be subclassed too, for example:
```csharp
var _result = await Commands.ExecuteAsync(context, argPos);
if (_result is MySubclassedResult result)
{
    var builder = new EmbedBuilder();
    for (var pair in result.Data)
    {
        builder.AddField(pair.Key, pair.Value, true);
    }
    await message.Channel.SendMessageAsync("", embed: builder);
}
else if (_result is RuntimeResult result)
{
    await message.Channel.SendMessageAsync(result.Reason);
}
else if (!_result.IsSuccess)
{
    // Previous error handling
}
```

khionu commented:
This was discussed in the Dev Guild.

khionu commented:
TL;DR, custom errors, or passing any data from the commands. Good for any PostCommand Logic

Joe4evr commented on src/Discord.Net.Commands/Builders/ModuleClassBuilder.cs:
```diff
@@ -249,7 +249,7 @@ private static bool IsValidModuleDefinition(TypeInfo typeInfo)
         {
             return methodInfo.IsDefined(typeof(CommandAttribute)) &&
-                   methodInfo.ReturnType == typeof(Task) &&
+                   (methodInfo.ReturnType == typeof(Task) || methodInfo.ReturnType == typeof(Task<RuntimeResult>)) &&
```
~~Since this PR (allegedly) still needs a discussion, I'm throwing this note in here: this check won't work *entirely* well when people want to return a `Task<CustomRuntimeResult>`.~~

Disregard this, I know where this decision is coming from now.

Joe4evr commented on src/Discord.Net.Commands/Results/RuntimeResult.cs:
```diff
@@ -0,0 +1,34 @@
+    [DebuggerDisplay(@"{DebuggerDisplay,nq}")]
+    public class RuntimeResult : IResult
+    {
+        private RuntimeResult(CommandError? error, string reason)
```
Should be `protected` for allowing subclassed RuntimeResults.

FiniteReality replied:
:upside_down_face: I'm sure I managed to subclass this when I tested it originally, good catch though.

FiniteReality committed:
Make RuntimeResult's ctor protected

FiniteReality committed:
Make RuntimeResult abstract

It never really made sense to have it instantiable in the first place,
frankly.

Joe4evr commented on src/Discord.Net.Commands/Info/CommandInfo.cs:
```diff
@@ -176,12 +175,22 @@ public Task<ExecuteResult> ExecuteAsync(ICommandContext context, ParseResult par
             {
-                await _action(context, args, services).ConfigureAwait(false);
+                var task = _action(context, args, services);
+                if (task is Task<IResult> returnTask)
```
I might be crazy, but this doesn't seem like it actually works. A quick test with [this code](https://gist.github.com/Joe4evr/1ba4e5abd7d7f122244cd7d681a45529) seems to print `Nope` every time.
It's possible to do `if (task is Task<RuntimeResult> resultTask)` instead, and not need the second check.

FiniteReality replied:
It does work, but probably not for users using command builders. Whoops.

The reason why I needed to return Task<IResult> is because of ModuleClassBuilder's Callback code: it returns a Task<IResult>, and unfortuantely due to contravariance not working the way I wanted it to, it would never trip the condition.
